Document/0x07a-Testing-Authentication-and-Session-Management.md

```diff
@@ -8,6 +8,65 @@ For all of the test cases below, it need to be investigated first what kind of a
+
+<img src="Images/Chapters/0x07a/abstract-oauth2-flow.png" width="350px"/>
+
+Here is a more detailed explanation of the steps in the diagram:
+
+1. The application requests authorization to access service resources from the user
+2. If the user authorized the request, the application receives an authorization grant
+3. The application requests an access token from the authorization server (API) by presenting authentication of its own identity, and the authorization grant
+4. If the application identity is authenticated and the authorization grant is valid, the authorization server (API) issues an access token to the application. Authorization is complete.
+5. The application requests the resource from the resource server (API) and presents the access token for authentication
+6. If the access token is valid, the resource server (API) serves the resource to the application
+
+Source: https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2
+
+The authorization grant might have different forms (explicit, implicit, etc), the access token might have a companion refresh token and might be used on different ways (e.g., as a bearer token). Please refer to https://tools.ietf.org/html/rfc6749 for further details.
+
+These are some of the common best practices for OAuth2 on native apps:
+
+User-agent:
+- Use an external user-agent (the browser) instead of an embedded user-agent (e.g. WebView or internal client user interface) to prevent End-User Credentials Phishing. However, doing this the app relies on the OS Keychain for server trust. This way it will not be possible to implement certificate pinning. A solution for this would be to restrict the embedded user-agent to only the relevant domain.
+- The user should have a way to verify visual trust mechanisms (e.g., Transport Layer Security (TLS) confirmation, web site mechanisms).
+- The client should validate the fully qualified domain name of the server to the public key presented by the server during connection establishment to prevent man-in-the-middle attacks.
+
+Type of grant:
+- Use code grant instead of implicit grant on native apps.
+- When using code grant, implement PKCE (Proof Key for Code Exchange) to protect the code grant. Make sure that the server also implements it.
+- The auth "code" should short-lived and only used inmidiately after receiving it. Make sure that they only reside on transient memory and are not stored or logged.
+
+Client secrets:
+- No shared secret should be used as proof of the client's identity as this could lead to client impersonation ("client_id" already serves this purpose). If for some reason they do use client secrets, be sure that they are stored in secure local storage.
+
+End-User credentials
+- The transmission of end-user credentials must be protected using transport-layer mechanisms such as TLS or any alternative that avoids the sending of plaintext credentials over the wire (e.g., Hash-based Message Authentication Code).
```

Comment:
I agree. At best you can use TLS and if you don't trust that, use a zero-knowledge-proof protocol such as SRP, but using an HMAC might get you into trouble (pass the hash-type of attacks, brute forcing it the moment the key is stale or not well protected, etc.)